### ISSUE TITLE
Rank deficiency of design matrix in multiple linear regression leads to cryptic error

### ISSUE THREAD

**Issue by sappelhoff** [raphaelvallat/pingouin#130]
Hi there, today I stumbled over an interesting error. Let me start with the example (I'll keep it close to the truth of what I actually do).

let's assume that:

- `n` are the number of trials in an experiment
- `y` are amplitudes in µV of [EEG](https://en.wikipedia.org/wiki/Electroencephalography) data at some channel/time bin (i.e., the voltage amplitude recorded at a specific EEG channel at a specific timepoint -> the same channel and timepoint per trial) 
- `X` is the design matrix --> imagine that the `n` trials in the experiment can come from 5 conditions (e.g., a number between 1 and 5 is shown as a stimulus on a screen), and each trial has a different condition ... the design matrix has as many columns as there are conditions. And in each row of the design matrix, the column corresponding to the condition of the current trial is switched to 1, the rest is 0. (and then, an intercept is added)

Ok, let's see the code:

```Python
import numpy as np
import pingouin
import statsmodels.api as sm

# make up some random data and a design matrix for multiple linear regression
n = 100 
y = np.random.randn(n)
X = np.vstack([np.random.permutation([0, 1, 0, 0, 0]) for i in range(n)])

# this will raise a ValueError
pingouin.linear_regression(X, y, add_intercept=True)

# Let's try ordinary least squares multiple regression in statsmodels instead
# This works, ... however there *is* a warning (see results.summary())
X_with_intercept = sm.add_constant(X)
model = sm.OLS(endog=y, exog=X_with_intercept)
results = model.fit()
results.summary()

# scikit learn multiple linear regression works as well
# ...
```

The ValueError is due to scipy's `lstsq` (which seems equivalent to `np.linalg.lstsq` to me), namely the rank of `X` (with added constant term) is one less than there are columns.

That in turn will lead to the residuals being unspecified, see the [docstr](https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html)

> Sums of residuals; squared Euclidean 2-norm for each column in b - a*x. **If the rank of a is < N or M <= N, this is an empty array**. If b is 1-dimensional, this is a (1,) shape array. Otherwise the shape is (K,).

Emphasis by me, in our case, the rank is < N.

I have several questions:

1. Why is the design matrix rank deficient in the way I set it up?
1. Why does statsmodels OLS work, but scipy, numpy, pingouin do not work?
1. Should we perhaps improve pingouin's error message?

regarding point 3, this is the current error:

> ValueError: operands could not be broadcast together with shapes (0,) (6,) 

it'd be more helpful to say something along these lines:

> Your design matrix is rank deficient, and residuals could not be meaningfully calculated ...

**Comment by raphaelvallat** [comment]
HI @sappelhoff,

Thanks for the detailed issue. Some first thoughts:

1) Your design matrix becomes rank deficient only when you add the intercept. It's because the intercept is always the sum of all the other columns (which is always 1), so your columns are not linearly independent. Pingouin works well when you pass `add_intercept=False`.

2) Sklearn only returns the rank given by scipy.linalg.lstsq, contrary to Pingouin it does not attempt to do any calculation with the empty array. I'll have to look into statsmodels's code, I'm guessing there must be a workaround to calculate the MSE when X is rank deficient. Or, perhaps they're not adding the constant when it could lead to a rank deficient matrix.

3) Agree! We should do something like:

```python
coef, ss_res, rank, _ = lstsq(Xw, yw)
if rank < p:
   warnings.warn("Design matrix is rank deficient. Bla bla bla.").
```

and then return NaN in the dataframe?

Thanks

**Comment by sappelhoff** [comment]
> Pingouin works well when you pass add_intercept=False

True, I have noticed that --> thanks for the explanation of the rank deficiency ... I didn't immediately see that.

Would it perhaps be a solution to z-score `y` and to not include a constant term (i.e., use `add_intercept=False`)?

>  I'll have to look into statsmodels's code, I'm guessing there must be a workaround to calculate the MSE when X is rank deficient. Or, perhaps they're not adding the constant when it could lead to a rank deficient matrix.

okay :thinking: yeah, would be interesting to know

> We should do something like:

yep, let me know if I should draft a PR :+1:

**Comment by raphaelvallat** [comment]
I'd prefer not directly changing the `y` variable, I think it's better to let the users know that there is an issue, and then they can make an informed decision. Please feel free to draft a PR. If you want to look at what statsmodels is doing under the good that would be fantastic as well, otherwise I'd have a look early next week.

Thanks!

**Comment by sappelhoff** [comment]
> I'd prefer not directly changing the y variable, I think it's better to let the users know that there is an issue, and then they can make an informed decision.

yeah, I absolutely agree ... that was more a question on what you'd think appropriate for my situation :-) 

okay, will draft a PR.

**Comment by sappelhoff** [comment]
was closes in #131 but somehow the "github closing automation" didn't work. perhaps because I don't have permissions. closing manually now.
